VideoTrackGenerator writer should get closed when its generator track (and all its clones) are stopped

youennf opened the pull request:
#### 8783423626ac3523165d5d1ec2794eabafba010c
<pre>
VideoTrackGenerator writer should get closed when its generator track (and all its clones) are stopped
<a href="https://rdar.apple.com/121835553">rdar://121835553</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=268279">https://bugs.webkit.org/show_bug.cgi?id=268279</a>

Reviewed by Jean-Yves Avenard.

Implement VideoTrackGenerator::Source::endProducingData to make sure to close VideoTrackGenerator writable when the source is stopped.
Introduce WritableStream::close to implement this.
The VideoTrackGenerator::Source keeps a WeakPtr of the WritableStream to call WritableStream::close.

We introduce closeWritableStreamIfPossible built-in to only close the writable stream if it is not already closed.

Add corresponding WPT tests.
We also fix a test in VideoTrackGenerator.worker.js which is about trying to write to a video track generator non VideoFrames.
We use AudioData, which cannot be created yet in Cocoa ports, hence the failure.

* LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker.js:
(make_audio_data):
(promise_test.async t):
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt: Added.
* Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp:
(WebCore::VideoTrackGenerator::create):
(WebCore::VideoTrackGenerator::Source::Source):
(WebCore::m_contextIdentifier):
(WebCore::VideoTrackGenerator::Source::endProducingData):
(WebCore::VideoTrackGenerator::Source::setWritable):
(WebCore::VideoTrackGenerator::Source::writeVideoFrame):
* Source/WebCore/Modules/mediastream/VideoTrackGenerator.h:
* Source/WebCore/Modules/streams/WritableStream.cpp:
(WebCore::WritableStream::closeIfPossible):
(WebCore::JSWritableStream::abort):
(WebCore::JSWritableStream::close):
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/streams/WritableStreamInternals.js:
(writableStreamCloseIfPossible):
* Source/WebCore/bindings/js/InternalWritableStream.cpp:
(WebCore::InternalWritableStream::abortForBindings):
(WebCore::InternalWritableStream::closeForBindings):
(WebCore::InternalWritableStream::closeIfPossible):
(WebCore::InternalWritableStream::abort): Deleted.
(WebCore::InternalWritableStream::close): Deleted.
* Source/WebCore/bindings/js/InternalWritableStream.h:

Canonical link: <a href="https://commits.webkit.org/273778@main">https://commits.webkit.org/273778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb729939f6178be2928b53fce514a3bf7c4d70b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36575 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15513 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39238 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32798 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37809 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17988 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12595 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31412 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37136 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13079 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32354 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11448 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11466 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32611 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40482 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32954 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37393 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11715 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9558 "Found 1 new test failure: inspector/canvas/requestShaderSource-webgl.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35500 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13399 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8308 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12139 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12600 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt:
```diff
@@ -2,6 +2,9 @@
 PASS Tests that creating a VideoTrackGenerator works as expected
 PASS Tests that VideoFrames are destroyed on write.
-PASS Mismatched frame and generator kind throws on write.
+FAIL Mismatched frame and generator kind throws on write. promise_test: Unhandled rejection with value: object "NotSupportedError: AudioData creation failed"
```
shouldn't the test be adapted so it doesn't fail here?

jyavenard commented on Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp:
```diff
@@ -122,13 +125,34 @@ Ref<MediaStreamTrack> VideoTrackGenerator::track()
+{
+    ASSERT(isMainThread());
+    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [weakThis = ThreadSafeWeakPtr { *this }] (auto&) {
+        auto protectedSource = weakThis.get();
```
`RefPtr protectedSource`

youennf replied:
get already returns a RefPtr, but yeah, this is almost coding style now

jyavenard commented on Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp:
```diff
@@ -122,13 +125,34 @@ Ref<MediaStreamTrack> VideoTrackGenerator::track()
+    ScriptExecutionContext::postTaskTo(m_contextIdentifier, [weakThis = ThreadSafeWeakPtr { *this }] (auto&) {
+        auto protectedSource = weakThis.get();
+        RefPtr writable = protectedSource ? protectedSource->m_writable.get() : nullptr;
+        if (writable)
```
this seems racy at a glance, we have m_writable written on the main thread but read on the event loop with no memory barriers (locks) in place.

If this is all main thread, can we add the `WTF_GUARDED_BY_CAPABILITY(mainThread)` required annotation for the m_writable member?

Will need to change the `ASSERT(isMainThread())` for `assertIsMainThread()`

youennf replied:
m_writable is only accessed in the worker thread (see the setter below), VideoTrackGenerator is only created in worker contexts.
The call to ScriptExecutionContext::postTaskTo makes sure we hop to the right thread to manipulate m_writable.
Also, m_writable is a RefCounted object so refing/derefing it in multiple threads should already be covered by asserts.

jyavenard replied:
My bad, I misread the code, I missed the `!`

youennf committed:
VideoTrackGenerator writer should get closed when its generator track (and all its clones) are stopped
rdar://121835553
https://bugs.webkit.org/show_bug.cgi?id=268279

Reviewed by Jean-Yves Avenard.

Implement VideoTrackGenerator::Source::endProducingData to make sure to close VideoTrackGenerator writable when the source is stopped.
Introduce WritableStream::close to implement this.
The VideoTrackGenerator::Source keeps a WeakPtr of the WritableStream to call WritableStream::close.

We introduce closeWritableStreamIfPossible built-in to only close the writable stream if it is not already closed.

Add corresponding WPT tests.
We also fix a test in VideoTrackGenerator.worker.js which is about trying to write to a video track generator non VideoFrames.
We use AudioData, which cannot be created yet in Cocoa ports, hence the failure.

* LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker.js:
(make_audio_data):
(promise_test.async t):
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/mediacapture-insertable-streams/VideoTrackGenerator.worker-expected.txt: Added.
* Source/WebCore/Modules/mediastream/VideoTrackGenerator.cpp:
(WebCore::VideoTrackGenerator::create):
(WebCore::VideoTrackGenerator::Source::Source):
(WebCore::m_contextIdentifier):
(WebCore::VideoTrackGenerator::Source::endProducingData):
(WebCore::VideoTrackGenerator::Source::setWritable):
(WebCore::VideoTrackGenerator::Source::writeVideoFrame):
* Source/WebCore/Modules/mediastream/VideoTrackGenerator.h:
* Source/WebCore/Modules/streams/WritableStream.cpp:
(WebCore::WritableStream::closeIfPossible):
(WebCore::JSWritableStream::abort):
(WebCore::JSWritableStream::close):
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/Modules/streams/WritableStreamInternals.js:
(writableStreamCloseIfPossible):
* Source/WebCore/bindings/js/InternalWritableStream.cpp:
(WebCore::InternalWritableStream::abortForBindings):
(WebCore::InternalWritableStream::closeForBindings):
(WebCore::InternalWritableStream::closeIfPossible):
(WebCore::InternalWritableStream::abort): Deleted.
(WebCore::InternalWritableStream::close): Deleted.
* Source/WebCore/bindings/js/InternalWritableStream.h:

Canonical link: https://commits.webkit.org/273778@main
